src/storage/src/monitor/hitmap.rs

```diff
@@ -0,0 +1,238 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+use risingwave_common::util::iter_util::ZipEqFast;
+
+#[derive(Debug, Clone)]
+pub struct Hitmap<const N: usize> {
+    /// For [`Hitmap`] is rarely access in multi-thread pattern,
+    /// the cons of false-sharing can be ignored.
+    data: Arc<[AtomicU64; N]>,
+}
+
+impl<const N: usize> Default for Hitmap<N> {
+    fn default() -> Self {
+        let data = [(); N].map(|_| AtomicU64::default());
+        let data = Arc::new(data);
+        Self { data }
+    }
+}
+
+impl<const N: usize> Hitmap<N> {
+    pub const fn bits() -> usize {
+        N * u64::BITS as usize
+    }
+
+    pub const fn bytes() -> usize {
+        N * 8
+    }
+
+    pub fn local(&self) -> LocalHitmap<N> {
+        LocalHitmap::new(self)
+    }
+
+    pub fn apply(&self, local: &LocalHitmap<N>) {
+        for (global, local) in self.data.iter().zip_eq_fast(local.data.iter()) {
+            global.fetch_or(*local, Ordering::Relaxed);
+        }
+    }
+
+    pub fn ones(&self) -> usize {
+        let mut res = 0;
+        for elem in &*self.data {
+            res += elem.load(Ordering::Relaxed).count_ones() as usize;
+        }
+        res
+    }
+
+    pub fn zeros(&self) -> usize {
+        Self::bits() - self.ones()
+    }
+
+    pub fn ratio(&self) -> f64 {
+        self.ones() as f64 / Self::bits() as f64
+    }
+
+    #[cfg(test)]
+    pub fn to_hex_vec(&self) -> Vec<String> {
+        use itertools::Itertools;
+        self.data
+            .iter()
+            .map(|elem| elem.load(Ordering::Relaxed))
+            .map(|v| format!("{v:016x}"))
+            .collect_vec()
+    }
+}
+
+#[derive(Debug)]
+pub struct LocalHitmap<const N: usize> {
+    owner: Hitmap<N>,
+    data: Box<[u64; N]>,
+}
+
+impl<const N: usize> LocalHitmap<N> {
+    pub const fn bits() -> usize {
+        N * u64::BITS as usize
+    }
+
+    pub const fn bytes() -> usize {
+        N * 8
+    }
+
+    pub fn new(owner: &Hitmap<N>) -> Self {
+        Self {
+            owner: owner.clone(),
+            data: Box::new([0; N]),
+        }
+    }
+
+    pub fn fill(&mut self, start_bit: usize, end_bit: usize) {
+        const MASK: usize = (1 << 6) - 1;
+
+        let end_bit = std::cmp::min(end_bit, Self::bits());
+
+        let head_bits = start_bit & MASK;
+        let tail_bits_rev = end_bit & MASK;
+
+        let head_elem = start_bit >> 6;
+        let tail_elem = end_bit >> 6;
+
+        for i in head_elem..=std::cmp::min(tail_elem, N - 1) {
```

Comment:
Is this cmp necessary if we have compared end_bit and `Self::bits`?

Reply:
For `end_bit` is exclusive, the comparison is needed here.